proposals/001-controlled-rollout.md

```diff
@@ -0,0 +1,115 @@
+# Controlled Rollout proposal
+
+## Why
+
+New features are great, everybody loves them. However, new features come by the hand of new code. New code can have bugs 
+or it can have performance regressions.
+
+Features aren't perfect for all users. Sometimes they are perfect for some users and not others. Sometimes they are 
+perfect for some hosts and not others.
+
+Rolling out a feature shouldn't always be a binary choice: enabled/disabled. In an ideal world, all features would be 
+enabled by default, everybody would love them all, and they would work flawlessly for all possible use cases.
+
+We are in the real world, though, which is not ideal. So we should give people running Fleet tools to rollout features 
+slowly, so that they can update infrastructure if needed, or only use a feature within the scope that is useful for 
+them.
+
+This is a proposal on how this tool could look like and work.
+
+## How
+
+We would create a new type of boolean value in our `AppConfig` called `RolloutBoolean`.
```

Comment:
I have typically seen this sort of thing called a feature flag. Our users may also be more familiar with that term than `RolloutBoolean`.

Reply:
Could be, feature flags in my book is a binary thing, though. This is a way of rolling out a feature so that eventually you enable for all if you want.

Reply:
Feature flags are not always just a binary off/on thing. For example, in [LaunchDarkly](https://docs.launchdarkly.com/home/flags/variations), feature flags can target a specific user, a subset of users, use conditional logic. They can also take a range of values (not just boolean, but also strings, numbers, etc)

Reply:
Note I've heard customers express concerns about turning on features for a bunch of hosts at once. They want to be able to roll out things like the software inventory queries slowly and monitor the performance impacts.

Reply:
> So we should give people running Fleet tools to ... use a feature within the scope that is useful for them.

The wording from the doc above suggests that this is intended as a config option for users, to control whether certain features are enabled

> it would be nicer to be able to say "enable for 10%" or "enable for hosts in this label"

I think a percentage rollout would be a great idea, especially for things that affect performance. 

> I have some concerns about being sure to properly handle the now per-host application of the features in all cases, it might get very tricky depending on the feature (e.g. you can't just check if the feature is enabled and if so, do a batch SELECT of things related to that feature - you have to check if each affected host has the feature enabled).

The best way I can think of implementing this right now is having a separate table you can join hosts with to check whether a feature is enabled eg `host_features`

Reply:
I intentionally left percentages out because of its added complexity. We can add it later on. If you want progressive rollout, either assign teams or specific hosts. If you want a range, we can make fleetctl auto generate the config based on a range, like @mna mentioned.

As for the wording, yup, it was left phrased like that so that this would serve both purposes: full AND partia rollouts. We can disagree on wether we want this, but the idea is: what's the simplest we could do that could serve the most use cases we've heard of?

Reply:
I can see percentages being useful, but I can also see customers requesting other ways to control this, I think we should ask around before settling. One crazy idea could be to control the target hosts via an osquery query

Reply:
> One crazy idea could be to control the target hosts via an osquery query

Not so crazy! This is basically (iiuc) being able to target labels :)